Add support for canceling CSS transitions

mrobinson opened the pull request:
This change adds support for canceling CSS transitions when a property is
no longer transitionable. Support for canceling and replacing CSS
transitions when the end value changes is still pending. This change
also takes advantage of updating the constellation message to fix a bug
where transition events could be sent for closed pipelines.

Fixes #15079.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15079.

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
Add support for canceling CSS transitions

This change adds support for canceling CSS transitions when a property
is no longer transitionable or when an element becomes styled with
display:none. Support for canceling and replacing CSS transitions when
the end value changes is still pending. This change also takes advantage
of updating the constellation message to fix a bug where transition
events could be sent for closed pipelines.

Fixes #15079.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 110ed9270880e17fb460128cc117284bc32b8938 with merge cee29f4174bc2f835a71ffad266e888888d10a33...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/PuQFgOHrRbWHxtptP_Z2iQ)

jdm commented:
Since this exposes a new DOM event, it seems surprising that no tests demonstrate a change in behaviour. Maybe add a test that shows that these changes are working?

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/23166.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#23166.

mrobinson commented:
It seems like the web-platform-tests have a few tests which detect transitioncancel when an element becomes `display: none` or when it's removed from the DOM, but it doesn't seem like there are any testing when a property is no longer transitionable. I've done two things:
 * Added a test which verifies this case.
 * Modified my change to handle the element becoming `display: none` causing one more WPT test to pass.

The rest of the tests that test `transitioncancel` events seem to rely on other features that we don't implement yet (such as `transitionrun` events or DOM removal). Hopefully these will start passing as we implement more events.

I'm not sure what the procedure is when adding new WPT tests, but I'm happy to work through that.

(the @servo-wpt-sync bot is really cool!)

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit a02b13dbf819b91c51bca916204268f49f34ae57 with merge 1114a85b71149b6ee97673b3edb5247993e2ed8a...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/XC5ENvnjRTWlAFmsYeTvvw)
State: approved= try=True

jdm commented:
```
{"path": "css/css-transitions/transitioncancel-002.html", "message": "Test-file line has a `console.*(...)` call", "lineno": 29, "rule": "CONSOLE"}
{"path": "css/css-transitions/transitioncancel-002.html", "message": "setTimeout used", "lineno": 39, "rule": "SET TIMEOUT"}
{"path": "css/css-transitions/transitioncancel-002.html", "message": "setTimeout used", "lineno": 44, "rule": "SET TIMEOUT"}
```

jdm commented on tests/wpt/web-platform-tests/css/css-transitions/transitioncancel-002.html:
```diff
@@ -0,0 +1,49 @@
+  getComputedStyle(div).backgroundColor;
+
+  // Remove the transitioning property from transition-property asynchronously.
+  setTimeout(() => { div.style.transitionProperty = 'none'; }, 0);
```
use t.step_timeout instead.

jdm commented on tests/wpt/web-platform-tests/css/css-transitions/transitioncancel-002.html:
```diff
@@ -0,0 +1,49 @@
+  await eventWatcher.wait_for('transitioncancel');
+
+  // There should be no transitionend events received.
+  await new Promise(resolve => setTimeout(resolve, 300));
```
t.step_timeout

jdm commented on tests/wpt/web-platform-tests/css/css-transitions/transitioncancel-002.html:
```diff
@@ -0,0 +1,49 @@
+  // Attach event listeners
+  const eventWatcher = new EventWatcher(t, div, ['transitioncancel']);
+  div.addEventListener('transitionend', t.step_func((event) => {
+    console.log(event);
```
remove this to satisfy the linter.

jdm reviewed (changes requested):
looks good! don't forget to run `mach update-manifest` after changing the test file.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#23166.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 88a1f6c has been approved by `jdm`

<!-- @bors-servo r=jdm 88a1f6cc6cb130abe164bd7bf2261af6c452627b -->

bors-servo commented:
:hourglass: Testing commit 88a1f6cc6cb130abe164bd7bf2261af6c452627b with merge 9ce6394899f29bee35675fa1a89810187c55754d...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/I8TQVMhDTmGi84AJnQPoCA)

jdm commented:
```
error[E0063]: missing field `cancelled_animations` in initializer of `style::context::SharedStyleContext<'_>`
   --> components/layout_thread_2020/lib.rs:605:28
    |
605 |             style_context: SharedStyleContext {
    |                            ^^^^^^^^^^^^^^^^^^ missing `cancelled_animations`
error: aborting due to previous error
```

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#23166.

mrobinson commented:
Thanks for the review. I've pushed a new version of this change which fixes the Layout 2020 build and also modifies the original change to only add entries for elements that actually do have canceled or running transitions, which is another issue I noticed after my most recent push.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 453b252 has been approved by `jdm`

<!-- @bors-servo r=jdm 453b252a655f82fd6405dea753f5a9363b414035 -->

bors-servo commented:
:hourglass: Testing commit 453b252a655f82fd6405dea753f5a9363b414035 with merge 61d5b477004a7235791fd77a13962b9ae5d290e7...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Pimzh5WqTt-O4cU63LTdwA)

jdm commented:
@bors-servo retry
* #25522

bors-servo commented:
:hourglass: Testing commit 453b252a655f82fd6405dea753f5a9363b414035 with merge ccdf2d7622b586cff6e16d8538d45519b92189eb...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/V_xYe33BQQ6HSzofW26tGw)

jdm commented:
@bors-servo retry
* #25513

bors-servo commented:
:hourglass: Testing commit 453b252a655f82fd6405dea753f5a9363b414035 with merge 610991bdfcc3b02b17f0dff9b744c9a8d1c3eaad...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/RtJc_KbrQwS_LdyzXPbtpw)
Approved by: jdm
Pushing 610991bdfcc3b02b17f0dff9b744c9a8d1c3eaad to master...
